Add compiler support for PostgreSQL "NOT VALID" constraints.

gilbsgilbs opened the pull request:
### Description
<!-- Describe your changes in detail -->

PostgreSQL supports declaring CHECK and FOREIGN KEY constraints as "NOT
VALID" which prevents the constraint from being validated on already
inserted rows, but still validates against any newly inserted or updated
rows.

This commit adds compiler support for this dialect option on
`CheckConstraint` and `ForeignKeyConstraint` classes. Although this
option doesn't make much sense on table creation, it doesn't seem to
cause any issue with Postgres and it is a preliminary step to have this
supported in Alembic during `ALTER TABLE`. Note that other constraints
types do not support this option.

See https://www.postgresql.org/docs/current/sql-altertable.html

See also #4825 which added supported for "NOT VALID" constraints
introsecpection.

Fixes #7600

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

gilbsgilbs committed:
Add compiler support for PostgreSQL "NOT VALID" constraints.

PostgreSQL supports declaring CHECK and FOREIGN KEY constraints as "NOT
VALID" which prevents the constraint from being validated on already
inserted rows, but still validates against any newly inserted or updated
rows.

This commit adds compiler support for this dialect option on
`CheckConstraint` and `ForeignKeyConstraint` classes. Although this
option doesn't make much sense on table creation, it doesn't seem to
cause any issue with Postgres and it is a preliminary step to have this
supported in Alembic during `ALTER TABLE`. Note that other constraints
types do not support this option.

See https://www.postgresql.org/docs/current/sql-altertable.html

See also #4825 which added supported for "NOT VALID" constraints
introsecpection.

Fixes: #7600

Signed-off-by: Gilbert Gilb's <gilbsgilbert@gmail.com>

zzzeek commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -2580,6 +2596,14 @@ def get_column_specification(self, column, **kwargs):
+            if not_valid:
+                return " NOT VALID"
+
+        return ""
```
it looks like NOT VALID doesn't actually change the functionality of the constraint so I guess it is safe to silently ignore for older versions; but in any case version 9.1 is super old at this point so it doesn't matter much anyway.

zzzeek replied:
our "normal support" is based at 9.6: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html  so I think just render the directive if it's asked for should be fine here

gilbsgilbs replied:
Fair enough. c2ed3e9b1c2e8510e139f40665bfcffcbf12e23d

zzzeek reviewed (commented):
not bad, looks pretty straightforward, thanks

gilbsgilbs committed:
Remove postgres version check for NOT VALID constraints.

- PostgreSQL 9.0 has reached EOL more than 6 years ago
- SQLAlchemy "normal support" is based at 9.6: https://docs.sqlalchemy.org/en/14/dialects/postgresql.html
- One can still not use the option and it will still work with older
  postgres version

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision c2ed3e9b1c2e8510e139f40665bfcffcbf12e23d of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change c2ed3e9b1c2e8510e139f40665bfcffcbf12e23d: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

needs a 1.4 backport + a changelog note



View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

gilbsgilbs commented:
@zzzeek I just added a changelog note for 1.4. Regarding the backport though, I'm not sure what you're expecting from me. Do you want me to open another PR or change my PR target branch to `rel_1_4`?

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

> Patch Set 1:
> 
> gilbsgilbs@github wrote:
> 
> @zzzeek I just added a changelog note for 1.4. Regarding the backport though, I'm not sure what you're expecting from me. Do you want me to open another PR or change my PR target branch to `rel_1_4`?

we can do it on this end, my note was out there for whomever wanted to finish this up, often @caselit or myself will go through and finish up PRs that need backports.  it's alittle tedious to do with pull requests so I wouldn't worry about it, thanks for writing the change note!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 78eecd55fd9fad07030d963f5fd6713c4af60e80 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 78eecd55fd9fad07030d963f5fd6713c4af60e80 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3557 has been **merged**. Congratulations! :)

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3568 has been **merged**. Congratulations! :)

gilbsgilbs commented:
Thanks a lot @zzzeek .
